src/tail/session.rs

```diff
@@ -0,0 +1,196 @@
+/// minutes, it will kill the tail session by removing the trace worker.
+impl Session {
+    pub async fn run(
+        target: Target,
+        user: GlobalUser,
+        mut shutdown_rx: Receiver<()>,
+        tx: Sender<()>,
+    ) -> Result<(), failure::Error> {
+        let client = http::cf_v4_api_client_async(&user, HttpApiClientConfig::default())?;
+
+        let url = get_tunnel_url(&mut shutdown_rx).await?;
+
+        let response = client
+            .request(&CreateTail {
+                account_identifier: &target.account_id,
+                script_name: &target.name,
+                params: CreateTailParams { url },
+            })
+            .await;
+
+        match response {
+            Ok(success) => {
+                eprintln!("Now prepared to stream logs.");
+
+                let tail_id = success.result.id;
+
+                // Loop indefinitely to send "heartbeat" to API and keep log streaming alive.
+                // This should loop forever until SIGINT is issued or Wrangler process is killed
+                // through other means.
+                let duration = Duration::from_millis(1000 * KEEP_ALIVE_INTERVAL);
+                let mut delay = delay_for(duration);
+
+                loop {
+                    match shutdown_rx.try_recv() {
+                        // this variant of the [TryRecvError](https://docs.rs/tokio/0.2.16/tokio/sync/oneshot/error/enum.TryRecvError.html)
+                        // occurs when the receiver listens for a message and the channel is empty;
+                        // in this case, it means that we have not received a shutdown command and
+                        // can continue with our task. The other variant would indicate that the
+                        // sender has been dropped, in which case we want to follow shut down as if
+                        // we received the signal.
+                        Err(TryRecvError::Empty) => {
+                            if delay.is_elapsed() {
+                                let heartbeat_result =
+                                    send_heartbeat(&target, &client, &tail_id).await;
+                                if heartbeat_result.is_err() {
+                                    return heartbeat_result;
+                                }
+                                delay = delay_for(duration);
+                            }
+                        }
+                        _ => {
+                            return Ok(());
+                        }
+                    }
+                }
+            }
+            Err(e) => {
+                tx.send(()).unwrap();
+                failure::bail!(http::format_error(e, Some(&tail_help)))
+            }
+        }
+    }
+}
+
+async fn get_tunnel_url(shutdown_rx: &mut Receiver<()>) -> Result<String, failure::Error> {
+    // regex for extracting url from cloudflared metrics port.
```

Comment:
```suggestion
    // extract hostname from cloudflared metrics port (e.g. https://ex.com:8080 -> ex.com)
```
Not sure but seemss this is getting the hostname not the URL? example could help readability

Reply:
agreed, though i intend to kill this regex and do something less dicey.

Reply:
For curiosity is the example I wrote accurate? Or what is this regex doing